### PR TITLE
new search

### PR DESCRIPTION
**Pull request by wsrvn**
improved index based search for maplist (/list), commandlist (/help) and songlist (in theory). tested on a 300000 maps set, it performs about 100 times faster (~60 seconds against ~0.5 seconds). on smaller sets, such as the 50000 maps server i tested it on, the search is essentially immediate (~0.03 seconds against ~5 seconds). ymmv of course, but i dont think it can be made much quicker.

other changes also include correct types for admin service commands to make them work with nicknametoplayer, which was apparently never used. now it is, and it allows for nickname to player (no way!) "translation".

how it works is: lets say theres a player with the nickname "michael" whose login is "73836382mich888ael", you could just type "//kick michael" and the controller will kick the correct login. obviously, if there are multiple players with similar nicknames, the closer match will be kicked but thats another story.

### CODE REVIEW

**Review by lythx — APPROVED** [review]
Good

File: src/Utils.ts [hunk 1]
```diff
@@ -422,25 +424,17 @@ export const Utils = {
...
+    matches.push(indices.map(a => players[a]))
     if (matches.length === 0) {
       return undefined
     }
-    const s = matches.sort((a, b): number => b.value - a.value)
-    if (s[0].value - s?.[1]?.value ?? 0 < options.minDifferenceBetweenMatches) {
-      return undefined
-    }
-    return s[0].player
+    return matches[0][0]
```
Comment (felacek) [inline src/Utils.ts:437]: I'm not sure this is the best solution. Maybe check if there are two people with the same nick and then ask the mod to clarify which one to kick?

File: src/Utils.ts [hunk 2]
```diff
@@ -640,3 +589,27 @@ function strVar(str: string, vars: { [name: string]: any }): string {
...
+function matchString(needle: string, haystack: string[]): number[] {
+  if (haystack.length === 0) { return [] }
+  haystack = ufuzzy.latinize(haystack)
+  const arr = []
+  const idxs = uf.filter(haystack, needle)
+  if (idxs != null && idxs.length > 0) {
+    const infoThresh = 1e4
+    if (idxs.length <= infoThresh) {
+      const info = uf.info(idxs, haystack, needle)
+      const order = uf.sort(info, haystack, needle)
```
Comment (felacek) [inline src/Utils.ts:608]: Maybe change variable names here as idxs isn't the best one.

**Review by felacek — COMMENTED** [review]
Not bad

File: src/Utils.ts (hunk 1 above)
Comment (wsrvn) [reply src/Utils.ts:437]: i dont see this ever becoming an issue unless there are 2 people with the same exact or very similar names. tbh, its their fault not ours in that case.

as for the confirmation message, that seems like something chatservice should do, not a random util...

File: src/Utils.ts (hunk 2 above)
Comment (wsrvn) [reply src/Utils.ts:608]: probably will rename them but tbh it matters not

**Pull request merged**